src/osd/scrubber/scrub_machine.h

```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
       const ReplicaReserveReq&,
       bool async_request);
 
+  /**
+   * the queued reservation request was granted by the async reserver.
+   * Notify the Primary.
+   * Returns 'false' if the reservation is not the last one to be received
+   * by this replica.
+   */
+  bool granted_by_reserver(const AsyncScrubResData& resevation);
+
   /// handle a 'release' from a primary
   void on_release(const ReplicaRelease& ev);
 
-  /// cancel the reserver request.
-  /// The 'failure' re 'log_failure' is logged if we are not reserved to
-  /// begin with.
+  /**
+   * cancel the reserver request.
+   * The 'failure' re 'log_failure' is logged if we are not reserved to
+   * begin with.
```

Comment:
The "we are not reserved to begin with" is meant to be impossible, which is why you are logging it.  For cases that shouldn't be possible, either be *very* explicit or remove it entirely.